lib/miq_cockpit.rb

```diff
@@ -0,0 +1,143 @@
+require 'uri'
+
+module MiqCockpit
+
+  class WS
+    DEFAULT_PORT = 9000
+
+    def self.url(opts, default_host)
+      if !opts
+        opts = {}
+      end
+
+      # running somewhere else
+      if (opts[:external_url])
+        url = URI.parse(opts[:external_url])
+      else
+        url = URI.parse("/cws/")
+        url.host = default_host
+        if (opts[:ssl])
+          url.scheme = "https"
+        else
+          url.scheme = "http"
+        end
+        url.port = opts[:port] || DEFAULT_PORT
+      end
+
+      if !url.path.ends_with?("/")
+        url.path = url.path + "/"
+      end
+
+      return url
+    end
+
+    def initialize(opts = {})
+      if !opts
+        opts = {}
+      end
+
+      @opts = opts
+      @config_dir = File.join(File.dirname(__FILE__), '../config')
+    end
+
+    def config_dir
+      @config_dir
+    end
+
+    def command (address)
+      cmd = ['/usr/libexec/cockpit-ws', "--port", @opts[:port] || "#{DEFAULT_PORT}"]
+      if (address)
+        cmd = cmd.push("--address", address)
+      end
+      if (!@opts[:ssl])
+        cmd = cmd.push("--no-tls")
+      end
+      return cmd
+    end
+
+    def save_config
+      Dir.mkdir(@config_dir) unless File.exists?(@config_dir)
+      Dir.mkdir(File.join(@config_dir, "cockpit")) unless File.exists?(File.join(@config_dir, "cockpit"))
+      fname = File.join(@config_dir, "cockpit", "cockpit.conf")
+      update()
+      File.open(fname, "w") { |f| f.write @config }
+      if @opts[:ssl]
+        setup_ssl()
+      end
+    end
+
+    def update
+      title = "ManageIQ Cockpit"
+
+      login_command =  File.join(File.dirname(__FILE__), 'cockpit', 'cockpit-auth-miq')
+      miq_url = @opts[:miq_url]
```

Comment:
Cockpit needs to know where to redirect back to when it needs a new auth token? If apache is in the mix, then a relative url is fine. But this can't be changed on a per-request basis. Is there a good way to get the public name of a miq server? Or if a config options like this the best I can  do?